pkg/utils/revision/revision_utils_test.go

```diff
@@ -0,0 +1,120 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package revision
+
+import (
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/utils/ptr"
+	leaderworkerset "sigs.k8s.io/lws/api/leaderworkerset/v1"
+)
+
+func TestApplyRevision(t *testing.T) {
+
+	lws := BuildLeaderWorkerSet("default")
+	revision, err := NewRevision(lws, 1, "")
+	if err != nil {
+		t.Fatal(err)
+	}
+	currentLws := lws.DeepCopy()
+
+	lws.Spec.LeaderWorkerTemplate.LeaderTemplate.Spec.Containers[0].Name = "update-name"
+	subdomainPolicy := leaderworkerset.SubdomainUniquePerReplica
+	lws.Spec.NetworkConfig = &leaderworkerset.NetworkConfig{
+		SubdomainPolicy: &subdomainPolicy,
+	}
+	restoredLws, err := ApplyRevision(lws, revision)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	restoredRevision, err := NewRevision(restoredLws, 2, "")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if !EqualRevision(revision, restoredRevision) {
+		t.Errorf("expected value %v, got %v", revision, restoredRevision)
+	}
+
+	if diff := cmp.Diff(currentLws.Spec.LeaderWorkerTemplate, restoredLws.Spec.LeaderWorkerTemplate); diff != "" {
+		t.Errorf("unexpected restored LeaderWorkerTemplate: %s", diff)
+	}
+
+	if diff := cmp.Diff(currentLws.Spec.NetworkConfig, restoredLws.Spec.NetworkConfig); diff != "" {
+		t.Errorf("NetworkConfig should be restored %s", diff)
+	}
+}
+
+func BuildLeaderWorkerSet(nsName string) *leaderworkerset.LeaderWorkerSet {
```

Comment:
Lets use the wrappers in https://github.com/kubernetes-sigs/lws/blob/main/test/testutils/wrappers.go

Reply:
Can't, otherwise it creates an import cycle. testutils imports the package revision in order to use GetPatch, and NewControllerRevision. So if I import testutils, the import cycle is testutils -> revision -> testutils. It was either recreating this function, or having a copy of GetPatch and NewController in testutils. Similar problem that we had with `tpu_test.go` https://github.com/kubernetes-sigs/lws/blob/main/pkg/utils/accelerators/tpu_test.go#L247

Reply:
We need to split the wrappers and validators into separate pkgs then. We can do that as a followup though.